### ISSUE TITLE
Recent commit causes new "Attempting to deserialize object on a CUDA device" exception

### ISSUE THREAD

**Issue by zanemountcastle** [ultralytics/yolov5#8001]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

PyTorch Hub

### Bug

I've been using YOLOv5 via PyTorch Hub for a few months now, but #7878 broke the deployment. I went from reliably running inference to getting the following error:

```
[ERROR] Exception: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.. Cache may be out of date, try `force_reload=True` or see https://github.com/ultralytics/yolov5/issues/36 for help.
```

The referenced issue was not helpful for solving the issue. What worked was reverting back to commit 
eb1217f3ac93fa906761216b16ae1b9435cbb850, which is just before the recent OpenVINO and MPS merges that seem to have caused the issue.

I'm now running YOLOv5 at commit eb1217f3ac93fa906761216b16ae1b9435cbb850 using the PyTorch Hub local feature and everything's back to normal. I'm not too familiar with the workings of YOLOv5 so I'm not able to spot where in the recent commits the bug is.

### Environment

- YOLO: YOLOv5 🚀 2022-5-26 Python-3.8.13 torch-1.10.2+cu102 CPU
- OS: AWS Linux
- Python 3.8.13

### Minimal Reproducible Example

Running the following: 
```
model = torch.hub.load('ultralytics/yolov5', 'custom', path='./custom_weights.pt', force_reload=True, device='cpu')
```
Causes the following error, even though this has been working for months in production systems:

```
[ERROR] Exception: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.. Cache may be out of date, try `force_reload=True` or see https://github.com/ultralytics/yolov5/issues/36 for help.
```

To fix it, I changed this to:
```
model = torch.hub.load('./yolov5', 'custom', path='./custom_weights.pt', force_reload=True, device='cpu', source='local')
```
Where `yolov5` is a clone of `ultralytics/yolov5` at commit eb1217f3ac93fa906761216b16ae1b9435cbb850.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@zanemountcastle user error. See [PyTorch Hub tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) for correct Usage examples.

**Comment by abdelaziz-mahdy** [comment]
i am having the same error during exporting the model 
running this note(best is a variable holding the location of the best.pt)
`!python export.py --weights {best} --include torchscript --imgsz 640 640 --optimize`

gets me this 
```
Traceback (most recent call last):
  File "export.py", line 607, in <module>
    main(opt)
  File "export.py", line 602, in main
    run(**vars(opt))
  File "/usr/local/lib/python3.7/dist-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "export.py", line 489, in run
    model = attempt_load(weights, device=device, inplace=True, fuse=True)  # load FP32 model
  File "/content/yolov5/models/experimental.py", line 80, in attempt_load
    ckpt = torch.load(attempt_download(w))
  File "/usr/local/lib/python3.7/dist-packages/torch/serialization.py", line 712, in load
    return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
  File "/usr/local/lib/python3.7/dist-packages/torch/serialization.py", line 1046, in _load
    result = unpickler.load()
  File "/usr/local/lib/python3.7/dist-packages/torch/serialization.py", line 1016, in persistent_load
    load_tensor(dtype, nbytes, key, _maybe_decode_ascii(location))
  File "/usr/local/lib/python3.7/dist-packages/torch/serialization.py", line 1001, in load_tensor
    wrap_storage=restore_location(storage, location),
  File "/usr/local/lib/python3.7/dist-packages/torch/serialization.py", line 176, in default_restore_location
    result = fn(storage, location)
  File "/usr/local/lib/python3.7/dist-packages/torch/serialization.py", line 152, in _cuda_deserialize
    device = validate_cuda_device(location)
  File "/usr/local/lib/python3.7/dist-packages/torch/serialization.py", line 136, in validate_cuda_device
    raise RuntimeError('Attempting to deserialize object on a CUDA '
RuntimeError: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.
```

**Comment by abdelaziz-mahdy** [comment]
also i am running on colab with gpu

**Comment by glenn-jocher** [comment]
@zezo357 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. Your issue is not reproducible:

<img width="1350" alt="Screen Shot 2022-05-27 at 6 25 13 PM" src="https://user-images.githubusercontent.com/26833433/170740021-ad9ce4d8-7afe-44e0-a0fc-d0e928c1b207.png">

We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by zanemountcastle** [comment]
I made a mistake in the original post. The code that causes the failure is:

```
model = torch.hub.load('ultralytics/yolov5', 'custom', path='./custom_weights.pt', force_reload=True, device='cpu')
```

I edited the original comment to reflect that.

I don't think this is simple user error since the exact same code but running an older version of YOLOv5 works and has been working for months. Pulling the latest code from `ultralytics/yolov5` using PyTorch Hub is what throws the error.

I'm running the code in a docker container running on AWS Lambda (which runs on AWS Linux), not Google Colab and has no GPU available.

The error is that it's reporting seems to be related to the lack of a GPU. I'm not familiar with Google Colab, so not sure if it's possible to reproduce the error there.

**Comment by abdelaziz-mahdy** [comment]
@glenn-jocher 

Just to help if someone had the same problem 

going to models folder
and editing experimental.py fixed the problem 
line 80 
from 
`ckpt = torch.load(attempt_download(w))`
to 
`ckpt = torch.load(attempt_download(w),map_location=torch.device(device))`

@zanemountcastle this may help you too

**Comment by zanemountcastle** [comment]
@zezo357 It seems like that could be the issue. I'll try those changes in a bit, but just looking at the changes to models/experimental.py at commit c215878f11d81808dfe4721795f0c105200e6601, it looks like `map_location` is no longer being passed under the hood. That seems to be causing YOLOv5 to fail at runtime on CPU machines since it is trying to run on a CUDA machine, which is causing the deserialization issue.

@glenn-jocher Is there a specific reason why `map_location` was removed from models/experimental.py in #7878? Could it be as simple as changing `ckpt = torch.load(attempt_download(w))` back to `ckpt = torch.load(attempt_download(w),map_location=torch.device(device))`? Happy to file a PR if that's the case.

**Comment by glenn-jocher** [comment]
@zanemountcastle @zezo357 I'm again unable to reproduce any issues with your command. I've run this just now myself in Colab and everything works correctly:

<img width="1302" alt="Screen Shot 2022-05-28 at 10 37 55 AM" src="https://user-images.githubusercontent.com/26833433/170817909-ab468aa1-b580-46a6-ba8c-ac893342e8ad.png">

We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by jlognn** [comment]
@zanemountcastle @zezo357 with your custom weights, did you wait for the model to completely finish training?  Or are you using a checkpoint taken during training?

**Comment by abdelaziz-mahdy** [comment]
@jlognn for me I did use a checkpoint.

**Comment by glenn-jocher** [comment]
> @zezo357 It seems like that could be the issue. I'll try those changes in a bit, but just looking at the changes to models/experimental.py at commit [c215878](https://github.com/ultralytics/yolov5/commit/c215878f11d81808dfe4721795f0c105200e6601), it looks like `map_location` is no longer being passed under the hood. That seems to be causing YOLOv5 to fail at runtime on CPU machines since it is trying to run on a CUDA machine, which is causing the deserialization issue.
> 
> @glenn-jocher Is there a specific reason why `map_location` was removed from models/experimental.py in #7878? Could it be as simple as changing `ckpt = torch.load(attempt_download(w))` back to `ckpt = torch.load(attempt_download(w),map_location=torch.device(device))`? Happy to file a PR if that's the case.

@zanemountcastle code changed for Apple MPS compatibility.

**Comment by glenn-jocher** [comment]
@zezo357 I tested a checkpoint without issues. If you are having issues please submit a **reproducible** example with exact code to reproduce.

```python
!python train.py --epochs 3 --save-period 1

import torch
model = torch.hub.load('ultralytics/yolov5', 'custom', 'runs/train/exp/weights/epoch1.pt')
print(model('https://ultralytics.com/images/bus.jpg'))
```

<img width="1371" alt="Screenshot 2022-05-30 at 13 04 12" src="https://user-images.githubusercontent.com/26833433/170979360-0240950f-c164-4621-a709-273461e4e759.png">

**Comment by abdelaziz-mahdy** [comment]
> @zezo357 I tested a checkpoint without issues. If you are having issues please submit a **reproducible** example with exact code to reproduce.
> 
> ```python
> !python train.py --epochs 3 --save-period 1
> 
> import torch
> model = torch.hub.load('ultralytics/yolov5', 'custom', 'runs/train/exp/weights/epoch1.pt')
> print(model('https://ultralytics.com/images/bus.jpg'))
> ```
> 
> <img width="1371" alt="Screenshot 2022-05-30 at 13 04 12" src="https://user-images.githubusercontent.com/26833433/170979360-0240950f-c164-4621-a709-273461e4e759.png">

I will try to replicate the error and give an example for it.

Thanks.

**Comment by abdelaziz-mahdy** [comment]
@glenn-jocher  Running 
`!python train.py --img 640 --batch 16 --epochs 30 --data coco128.yaml --weights yolov5s.pt --cache`
and stop while training (like spoch 3)
![image](https://user-images.githubusercontent.com/25157308/170989005-9564ff00-e6fc-428b-947c-137714cb0f6f.png)

then running 
`!python export.py --weights /content/yolov5/runs/train/exp2/weights/best.pt --include torchscript --imgsz 640 640 --optimize`


gives this error 

![image](https://user-images.githubusercontent.com/25157308/170988945-1babafe0-bc78-44dd-940f-73739a74e412.png)

**Comment by glenn-jocher** [comment]
@zezo357 thank you. I am able to reproduce. Will investigate.

**Comment by glenn-jocher** [comment]
@zezo357 @jlognn @zanemountcastle good news 😃! Your original issue may now be fixed ✅ in PR #8051. This PR reintroduces the `map_location` argument in attempt_load() to how it was prior to the Apple MPS PR.

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by spacewalk01** [comment]
Got same problem any solution? I trained a custom model and tried to export it to onnx but got the following error:

```
Traceback (most recent call last):
  File "export.py", line 607, in <module>
    main(opt)
  File "export.py", line 602, in main
    run(**vars(opt))
  File "F:\Software\anaconda\envs\pytorch\lib\site-packages\torch\autograd\grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "export.py", line 489, in run
    model = attempt_load(weights, device=device, inplace=True, fuse=True)  # load FP32 model
  File "F:\Projects\waterline-detection-yolov5\yolov5\models\experimental.py", line 80, in attempt_load
    ckpt = torch.load(attempt_download(w))
  File "F:\Software\anaconda\envs\pytorch\lib\site-packages\torch\serialization.py", line 712, in load
    return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
  File "F:\Software\anaconda\envs\pytorch\lib\site-packages\torch\serialization.py", line 1049, in _load
    result = unpickler.load()
  File "F:\Software\anaconda\envs\pytorch\lib\site-packages\torch\serialization.py", line 1019, in persistent_load
    load_tensor(dtype, nbytes, key, _maybe_decode_ascii(location))
  File "F:\Software\anaconda\envs\pytorch\lib\site-packages\torch\serialization.py", line 1001, in load_tensor
    wrap_storage=restore_location(storage, location),
  File "F:\Software\anaconda\envs\pytorch\lib\site-packages\torch\serialization.py", line 175, in default_restore_location
    result = fn(storage, location)
  File "F:\Software\anaconda\envs\pytorch\lib\site-packages\torch\serialization.py", line 152, in _cuda_deserialize
    device = validate_cuda_device(location)
  File "F:\Software\anaconda\envs\pytorch\lib\site-packages\torch\serialization.py", line 136, in validate_cuda_device
    raise RuntimeError('Attempting to deserialize object on a CUDA '
RuntimeError: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.
```
